SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AutodiffClosureSpecialization.swift

```diff
@@ -0,0 +1,625 @@
+                        handledFuncs.insert(faiCallee)
+                        if inner(faiCallee, fai.calleeArgumentIndex(of: use)!, &handledFuncs) {
+                            return true
+                        }
+                    }
+                }
+            }
+        }
+
+        return false
+    }
+
+    // Limit the number of recursive calls to not go into 
+    // exponential behavior in corner cases.
+    let recursionBudget = 8
+    var handledFuncs: Set<Function> = []
+    return inner(callee, index, &handledFuncs)
+}
+
+/// Marks any intermediate reabstracted closures corresponding to given closure
+/// passed in at a callsite as used. We do not want to look at reabstraction closures 
+/// twice as they are covered as part of the final closure that we're passing in at the  
+/// callsite.
+func markReabstractionPartialApplyAsUsed(firstClosure: Value, currentClosure: Value, usedReabstractionClosures: inout InstructionSet) {
+    if currentClosure == firstClosure {
+        return
+    }
+
+    if let pai = currentClosure as? PartialApplyInst {
+        usedReabstractionClosures.insert(pai)
+        return markReabstractionPartialApplyAsUsed(firstClosure: firstClosure, currentClosure: pai.callee, usedReabstractionClosures: &usedReabstractionClosures)
+    }
+
+    if let cvt = currentClosure as? ConvertFunctionInst {
+        usedReabstractionClosures.insert(cvt)
+        return markReabstractionPartialApplyAsUsed(firstClosure: firstClosure, currentClosure: cvt.fromFunction, usedReabstractionClosures: &usedReabstractionClosures)
+    }
+
+    if let cvt = currentClosure as? ConvertEscapeToNoEscapeInst {
+        usedReabstractionClosures.insert(cvt)
+        return markReabstractionPartialApplyAsUsed(firstClosure: firstClosure, currentClosure: cvt.fromFunction, usedReabstractionClosures: &usedReabstractionClosures)
+    }
+
+    if let mdi = currentClosure as? MarkDependenceInst {
+        usedReabstractionClosures.insert(mdi)
+        return markReabstractionPartialApplyAsUsed(firstClosure: firstClosure, currentClosure: mdi.value, usedReabstractionClosures: &usedReabstractionClosures)
+    }   
+
+    fatalError("Unexpected instruction used during closure reabstraction: \(currentClosure)")
+}
+
+// ===================== Types ===================== //
+
+/// Represents all the information required to
+/// represent a closure in isolation, i.e., outside of
+/// a callsite context where the closure may be getting
+/// passed as an argument.
+///
+/// Composed with other information inside a `ClosureArgDescriptor`
+/// to represent a closure as an argument at a callsite.
+class ClosureInfo {
+    let closure: SingleValueInstruction
+    var lifetimeFrontier: [Instruction]?
+
+    init(closure: SingleValueInstruction) {
+        self.closure = closure
+        self.lifetimeFrontier = nil
+    }
+
+    public func setLifetimeFrontier(_ frontier: [Instruction]) {
+        if self.lifetimeFrontier == nil {
```

Comment:
That's scary. What if `setLifetimeFrontier` is called twice? Is it intention that the frontier is ignored the second time?

Reply:
> Is it intention that the frontier is ignored the second time?
That is the intention here. It also ties to the question you've asked about `ClosureInfo` being a class. 

We can avoid exposing this method and the current constructor outside this file or use type-state pattern in some capacity. But the final solution, I think, will depend on whether or not `ClosureInfo` is a class.

Reply:
This method has been removed in the latest revision.